In [1]:
import numpy as np
import seaborn as sns
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import train_test_split

In [2]:
df=pd.read_csv('wafer_23012020_041211.csv')
df.head()

Unnamed: 0  Sensor-1  Sensor-2   Sensor-3   Sensor-4  Sensor-5  Sensor-6  \
0  Wafer-801   2968.33   2476.58  2216.7333  1748.0885    1.1127     100.0   
1  Wafer-802   2961.04   2506.43  2170.0666  1364.5157    1.5447     100.0   
2  Wafer-803   3072.03   2500.68  2205.7445  1363.1048    1.0518     100.0   
3  Wafer-804   3021.83   2419.83  2205.7445  1363.1048    1.0518     100.0   
4  Wafer-805   3006.95   2435.34  2189.8111  1084.6502    1.1993     100.0   

   Sensor-7  Sensor-8  Sensor-9  ...  Sensor-582  Sensor-583  Sensor-584  \
0   97.5822    0.1242    1.5300  ...         NaN      0.5004      0.0120   
1   96.7700    0.1230    1.3953  ...         NaN      0.4994      0.0115   
2  101.8644    0.1220    1.3896  ...         NaN      0.4987      0.0118   
3  101.8644    0.1220    1.4108  ...         NaN      0.4934      0.0123   
4  104.8856    0.1234    1.5094  ...         NaN      0.4987      0.0145   

   Sensor-585  Sensor-586  Sensor-587  Sensor-588  Sensor-589  Sensor-590  \
0      0.0033      2.4069      0.0545      0.0184      0.0055     33.7876   
1      0.0031      2.3020      0.0545      0.0184      0.0055     33.7876   
2      0.0036      2.3719      0.0545      0.0184      0.0055     33.7876   
3      0.0040      2.4923      0.0545      0.0184      0.0055     33.7876   
4      0.0041      2.8991      0.0545      0.0184      0.0055     33.7876   

   Good/Bad  
0        -1  
1         1  
2        -1  
3        -1  
4        -1  

[5 rows x 592 columns]

In [3]:
df.shape

(100, 592)

In [4]:
df.isna().sum().sum()

2306

In [5]:
df.drop(['Unnamed: 0','Good/Bad'],axis=1).iloc[:100].to_csv('test.csv',index=False)

In [6]:
df.isna().sum().sum()

2306

In [7]:
df,df_test=train_test_split(df,random_state=42,test_size=0.2)

In [8]:
df.isna().sum().sum()

1822

In [9]:
df.shape

(80, 592)

In [10]:
df.isna().sum().sum()/(df.shape[0]*(df.shape[1]-1))*100

3.8536379018612523

## Approximatelly 4% are nans

In [11]:
df['Good/Bad'].value_counts()

Good/Bad
-1    74
 1     6
Name: count, dtype: int64

In [12]:
df.shape

(80, 592)

## Therefore unbalanced dataset

### The dataset with 0 standard deviation needs to be drooped

In [13]:
# Deleting the columns which have std==0
def get_col_with_0_std(df: pd.DataFrame):
    col_to_drop=[]
    num_col=[column for column in df.columns if df[column].dtype !='O']
    for i in num_col:
        if df[i].std()==0:
            col_to_drop.append(i)
    return col_to_drop

#Deleting the columns which has nans > some threshold
def get_redundant_cols(df: pd.DataFrame, threshold):
    cols_missing_ratio=df.isna().sum().div(df.shape[0])
    col_to_drop=list(cols_missing_ratio[cols_missing_ratio>threshold].index)
    return col_to_drop

In [14]:
col1=get_col_with_0_std(df)
len(col1)

122

In [15]:
col2=get_redundant_cols(df,0.7)
col2

['Sensor-158', 'Sensor-159', 'Sensor-293', 'Sensor-294']

In [16]:
total_col=col1+col2
len(total_col)

126

In [17]:
total_col.append('Unnamed: 0')
len(total_col)

127

In [18]:
total_col

['Sensor-6',
 'Sensor-14',
 'Sensor-43',
 'Sensor-50',
 'Sensor-53',
 'Sensor-70',
 'Sensor-75',
 'Sensor-98',
 'Sensor-142',
 'Sensor-150',
 'Sensor-179',
 'Sensor-180',
 'Sensor-187',
 'Sensor-190',
 'Sensor-191',
 'Sensor-192',
 'Sensor-193',
 'Sensor-194',
 'Sensor-195',
 'Sensor-207',
 'Sensor-210',
 'Sensor-227',
 'Sensor-230',
 'Sensor-231',
 'Sensor-232',
 'Sensor-233',
 'Sensor-234',
 'Sensor-235',
 'Sensor-236',
 'Sensor-237',
 'Sensor-238',
 'Sensor-241',
 'Sensor-242',
 'Sensor-243',
 'Sensor-244',
 'Sensor-257',
 'Sensor-258',
 'Sensor-259',
 'Sensor-260',
 'Sensor-261',
 'Sensor-262',
 'Sensor-263',
 'Sensor-264',
 'Sensor-265',
 'Sensor-266',
 'Sensor-267',
 'Sensor-277',
 'Sensor-285',
 'Sensor-314',
 'Sensor-315',
 'Sensor-316',
 'Sensor-323',
 'Sensor-326',
 'Sensor-327',
 'Sensor-328',
 'Sensor-329',
 'Sensor-330',
 'Sensor-331',
 'Sensor-343',
 'Sensor-348',
 'Sensor-365',
 'Sensor-370',
 'Sensor-371',
 'Sensor-372',
 'Sensor-373',
 'Sensor-374',
 'Sensor-375',
 'Se

### Therefore 126 columns to drop in total

In [19]:
x,y=df.drop(total_col,axis=1),df[['Good/Bad']]

In [20]:
x

Sensor-1  Sensor-2   Sensor-3   Sensor-4  Sensor-5  Sensor-7  Sensor-8  \
55       NaN   2532.45  2191.1333  2197.6570    1.1569   89.7222    0.1251   
88   3221.21   2391.20  2189.9667  1046.6212    0.8662  102.3622    0.1208   
26   2951.85   2525.00  2189.5777  1320.3197    1.3459  100.7744    0.1234   
42   2982.07   2447.06  2199.6334  1242.8420    1.4083   99.2178    0.1221   
69   3058.08   2524.60  2192.3778  1110.5453    0.8147   99.2922    0.1226   
..       ...       ...        ...        ...       ...       ...       ...   
60   3071.05   2642.15  2200.9889  1054.5240    1.3830  100.1800    0.1201   
71   3043.18   2545.53  2192.3778  1110.5453    0.8147   99.2922    0.1226   
14   3001.26   2519.92  2224.6778  1308.6479    1.3907  101.1333    0.1208   
92   3007.00   2572.62  2213.2111  2070.7147    1.9705   87.7411    0.1232   
51   3078.77   2533.04  2187.4111  1942.3069    1.1864   88.0911    0.1245   

    Sensor-9  Sensor-10  Sensor-11  ...  Sensor-582  Sensor-583  Sensor-584  \
55    1.5762     0.0028    -0.0066  ...         NaN      0.4936      0.0113   
88    1.4756    -0.0025     0.0025  ...         NaN      0.4940      0.0123   
26    1.5590    -0.0032     0.0135  ...     53.8577      0.5025      0.0178   
42    1.4542     0.0142    -0.0064  ...         NaN      0.4993      0.0151   
69    1.4958     0.0004     0.0037  ...     24.6547      0.4974      0.0171   
..       ...        ...        ...  ...         ...         ...         ...   
60    1.4532     0.0049    -0.0048  ...         NaN      0.4973      0.0139   
71    1.3824    -0.0001    -0.0050  ...         NaN      0.4989      0.0131   
14    1.5172    -0.0135     0.0070  ...     48.4818      0.4959      0.0142   
92    1.4446    -0.0050    -0.0007  ...         NaN      0.4987      0.0172   
51    1.4500    -0.0053     0.0005  ...     46.7092      0.4968      0.0202   

    Sensor-585  Sensor-586  Sensor-587  Sensor-588  Sensor-589  Sensor-590  \
55      0.0033      2.2874      0.0133      0.0139      0.0038    104.3034   
88      0.0033      2.4860      0.0280      0.0078      0.0022     27.7601   
26      0.0045      3.5361      0.0286      0.0154      0.0056     53.8577   
42      0.0038      3.0214      0.0117      0.0262      0.0089    223.1018   
69      0.0040      3.4352      0.0218      0.0054      0.0020     24.6547   
..         ...         ...         ...         ...         ...         ...   
60      0.0039      2.7851      0.0122      0.0131      0.0039    107.5257   
71      0.0036      2.6253      0.0218      0.0054      0.0020     24.6547   
14      0.0037      2.8609      0.0278      0.0135      0.0042     48.4818   
92      0.0041      3.4417      0.0195      0.0149      0.0047     76.0035   
51      0.0052      4.0733      0.0262      0.0123      0.0033     46.7092   

    Good/Bad  
55        -1  
88        -1  
26        -1  
42        -1  
69        -1  
..       ...  
60        -1  
71        -1  
14        -1  
92        -1  
51        -1  

[80 rows x 465 columns]

### We also need to drop this Unnamed column its useless hence total 127 columns to drop

In [21]:
x.shape

(80, 465)

In [22]:
y

Good/Bad
55        -1
88        -1
26        -1
42        -1
69        -1
..       ...
60        -1
71        -1
14        -1
92        -1
51        -1

[80 rows x 1 columns]

In [23]:
y.shape

(80, 1)

# Data Transformation

In [24]:
from sklearn.impute import KNNImputer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import RobustScaler   #Using robust scaler as to handle outliers

In [25]:
imputer=KNNImputer(n_neighbors=3)
pre_processing_pipeline=Pipeline(
    steps=[('Imputer',imputer),
           ('Scaler',RobustScaler())]
)
pre_processing_pipeline

Pipeline(steps=[('Imputer', KNNImputer(n_neighbors=3)),
                ('Scaler', RobustScaler())])

In [26]:
x_trans=pre_processing_pipeline.fit_transform(x)
x_trans.shape

(80, 465)

In [1]:
x_trans

NameError: name 'x_trans' is not defined

In [1]:
import os
from datetime import datetime
os.getcwd()

'c:\\Users\\Dell\\Desktop\\PROJECT\\Wafer_fault\\Notebook'

In [5]:
LOG_FILE = f"{datetime.now().strftime('%m_%d_%Y_%H_%M_%S')}.log"

logs_path = os.path.join(os.getcwd(), "logs", LOG_FILE)

os.makedirs(logs_path, exist_ok=True)

LOG_FILE_PATH = os.path.join(logs_path,LOG_FILE)

In [6]:
LOG_FILE_PATH

'c:\\Users\\Dell\\Desktop\\PROJECT\\Wafer_fault\\Notebook\\logs\\06_29_2023_13_22_23.log\\06_29_2023_13_22_23.log'

In [4]:
import logging
logging.basicConfig(filename=LOG_FILE_PATH,format='%(asctime)s %(name)s %(levelname)s %(message)s',level=logging.INFO)

In [5]:
logging.error("HELP")

In [6]:
logging.shutdown()

In [7]:
logging.error("HELP")

In [8]:
logging.shutdown()

In [7]:
import logging
logging.basicConfig(filename=LOG_FILE_PATH,format='%(asctime)s %(name)s %(levelname)s %(message)s',level=logging.INFO)

In [12]:
logging.info("WOW")